### PR TITLE
[WALL-14932]: Refactor Android SDK to support multiple wallet connections

### PR DESCRIPTION
**Pull request by AMITGOELNY**
### _Summary_

Refactor Android SDK to support multiple wallet connections

* Added configure method for client to pass in callback and context
* Added `getClient` method to retrieve and manage SDK instances by `host` Wallet
* Created `MobileWalletProvider`
* Created interfaces for `KeyManager` and `TaskManager`
* Added Wallet Picker to Android example app

### _How did you test your changes?_

https://user-images.githubusercontent.com/4512097/200977387-42697017-d77f-4bfb-a677-bc0cff620a9b.mov

### CODE REVIEW

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt [hunk 1]
```diff
@@ -27,43 +29,44 @@ import java.util.UUID
 const val CBW_PACKAGE_NAME = "org.toshi"
 private const val CBW_SCHEME = "cbwallet://wsegue"
 
-class CoinbaseWalletSDK(
-    domain: Uri,
-    private val appContext: Context,
+class CoinbaseWalletSDK internal constructor(
     private val hostPackageName: String = CBW_PACKAGE_NAME,
+    private val scheme: String = CBW_SCHEME,
```
Comment (vishnumad) [inline android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt]: Maybe defaults can be removed now that this constructor is internal? I don't think there's a case where we want to default to CB Wallet anymore.

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt [hunk 2]
```diff
@@ -238,4 +239,32 @@ class CoinbaseWalletSDK(
...
+            return instances[wallet.url] ?: CoinbaseWalletSDK(
+                openIntent = openIntent,
+                hostPackageName = wallet.packageName,
+                scheme = wallet.url,
+                keyManager = KeyManager(context, wallet.packageName),
+                taskManager = TaskManager()
+            )
+        }
+
+        fun handleResponse2(uri: Uri) {
```
Comment (vishnumad) [inline android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt]: Rename to `handleResponse`

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt [hunk 3]
```diff
@@ -238,4 +239,32 @@ class CoinbaseWalletSDK(
...
     }
+
+    companion object {
+        private val instances: MutableMap<String, CoinbaseWalletSDK> = mutableMapOf()
+
+        private lateinit var callback: Uri
+
+        private lateinit var context: Application
+
+        fun configure(callback: Uri, context: Application) {
```
Comment (vishnumad) [inline android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt]: Maybe `callback` here should be something else since it's not being used on Android for calling back. `domain`?

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt [hunk 4]
```diff
@@ -238,4 +239,32 @@ class CoinbaseWalletSDK(
...
+                hostPackageName = wallet.packageName,
+                scheme = wallet.url,
+                keyManager = KeyManager(context, wallet.packageName),
+                taskManager = TaskManager()
+            )
+        }
+
+        fun handleResponse2(uri: Uri) {
+            val callback = checkNotNull(MessageConverter.getCallbackFromResponse(uri)) { "Callback not found" }
+            instances[callback]?.handleResponse(uri)
```
Comment (vishnumad) [inline android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt]: Let's return the Boolean value that the `handleResponse` function on the SDK instance returns. This way the dapps can know if the request was handled by the sdk.

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/MobileWalletProvider.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    suspend fun getWallets(): List<Wallet>
+}
+
+@Serializable
+data class Wallet(
+    val name: String,
+    val iconUrl: String,
+    val packageName: String,
+    val url: String,
+    val isInstalled: ((Boolean) -> Unit)? = null
```
Comment (vishnumad) [inline android/walletsdk/src/main/java/com/coinbase/android/nativesdk/MobileWalletProvider.kt]: `isInstalled` lambda should probably return a boolean. Is there a reason why this is nullable?

```kt
val isInstalled: () -> Boolean
```

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/MobileWalletProvider.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        iconUrl = "https://play-lh.googleusercontent.com/wrgUujbq5kbn4Wd4tzyhQnxOXkjiGqq39N4zBvCHmxpIiKcZw_Pb065KTWWlnoejsg",
+        packageName = "org.toshi",
+        url = "cbwallet://wsegue"
+    )
+
+    val coinbaseRetail = Wallet(
+        name = "Coinbase",
+        iconUrl = "https://play-lh.googleusercontent.com/PjoJoG27miSglVBXoXrxBSLveV6e3EeBPpNY55aiUUBM9Q1RCETKCOqdOkX2ZydqVf0",
+        packageName = "org.toshi.debugger",
+        url = "cbwallet22://wsegue"
```
Comment (vishnumad) [inline android/walletsdk/src/main/java/com/coinbase/android/nativesdk/MobileWalletProvider.kt]: Can we update these to the actual Retail app's values or add a TODO: to update these in the future?

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/key/KeyManager.kt [hunk 7]
```diff
@@ -53,15 +53,19 @@ internal class KeyManager(appContext: Context, host: String) {
...
-            // Generate new key pair and save to encrypted storage
-            val kp = EllipticCurves.generateKeyPair(EllipticCurves.CurveType.NIST_P256)
-            encryptedStore.saveKeyPair(alias, kp)
-            kp
-        }
+        // If no keypair, generate new key pair and save to encrypted storage
+        return encryptedStore.getKeyPair(alias)
+            ?: EllipticCurves.generateKeyPair(EllipticCurves.CurveType.NIST_P256).also {
+                encryptedStore.saveKeyPair(alias, it)
+            }
```
Comment (vishnumad) [inline android/walletsdk/src/main/java/com/coinbase/android/nativesdk/key/KeyManager.kt]: Nit: Maybe this can be a separate function. I find multiline ternaries a bit hard to parse
Something like `return encryptedStore.getKeyPair(alias) ?: generateKeyPair(alias)`

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/message/MessageConverter.kt [hunk 8]
```diff
@@ -78,6 +78,13 @@ object MessageConverter {
         return JSON.decodeFromJsonElement(encryptedSerializer, messageJson)
     }
 
+    fun getCallbackFromResponse(uri: Uri): String? {
+        val encodedMessage = requireNotNull(uri.getQueryParameter("p"))
+        val messageJsonString = String(Base64.decode(encodedMessage))
+        val messageJson = JSON.parseToJsonElement(messageJsonString)
+        return messageJson.jsonObject["callbackUrl"]?.jsonPrimitive?.content
+    }
```
Comment (vishnumad) [inline android/walletsdk/src/main/java/com/coinbase/android/nativesdk/message/MessageConverter.kt]: Note: We might want to create a `decodeWithoutDecrypting` function similar to what exists on iOS. This would be useful on the host library as well. Probably will require changes to the serialization/deserialization logic. Let's discuss async.

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt (hunk 1 above)
Comment (AMITGOELNY) [reply android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt]: removed defaults

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt (hunk 2 above)
Comment (AMITGOELNY) [reply android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt]: renamed

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt (hunk 3 above)
Comment (AMITGOELNY) [reply android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt]: renamed to domain

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt (hunk 4 above)
Comment (AMITGOELNY) [reply android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt]: updated

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/MobileWalletProvider.kt (hunk 5 above)
Comment (AMITGOELNY) [reply android/walletsdk/src/main/java/com/coinbase/android/nativesdk/MobileWalletProvider.kt]: moved to ext method.

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/key/KeyManager.kt (hunk 7 above)
Comment (AMITGOELNY) [reply android/walletsdk/src/main/java/com/coinbase/android/nativesdk/key/KeyManager.kt]: updated, but this is still binary eval, not ternary.  created generateKeyPair as its own method

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/message/MessageConverter.kt (hunk 8 above)
Comment (AMITGOELNY) [reply android/walletsdk/src/main/java/com/coinbase/android/nativesdk/message/MessageConverter.kt]: Let's follow up on this accordingly

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt [hunk 9]
```diff
@@ -238,4 +229,38 @@ class CoinbaseWalletSDK(
...
+                    .appendPath("wsegue")
+                    .build()
+            } else {
+                domain
+            }
+            this.context = context
+        }
+
+        fun getClient(wallet: Wallet, openIntent: (Intent) -> Unit): CoinbaseWalletSDK {
+            return instances[wallet.url] ?: CoinbaseWalletSDK(
```
Comment (vishnumad) [inline android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt:253]: If there's already an instance of the CoinbaseWalletSDK in the map, the passed in `openIntent` lambda will not be called when a new request is created, right? Do you think this can cause problems? 
Maybe the `openIntent` callback can be moved into a separate function where it's saved as a static instance.

**Review by vishnumad — DISMISSED** [review]

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/MobileWalletProvider.kt [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import kotlinx.serialization.Serializable
+
+object MobileWalletProvider : MobileWalletProviderInterface {
+
+    private val wallets = listOf(DefaultWallets.coinbaseWallet, DefaultWallets.coinbaseRetail)
+
+    override suspend fun getWallets(): List<Wallet> = wallets
+}
+
+internal interface MobileWalletProviderInterface {
```
Comment (bangtoven) [inline android/walletsdk/src/main/java/com/coinbase/android/nativesdk/MobileWalletProvider.kt]: image

File: android/example/src/main/java/com/coinbase/android/beta/ui/WalletProviderBottomSheetFragment.kt [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+import androidx.lifecycle.lifecycleScope
+import com.coinbase.android.beta.R
+import com.coinbase.android.beta.databinding.BottomSheetBinding
+import com.coinbase.android.beta.ui.adapter.WalletAdapter
+import com.coinbase.android.beta.utils.SpacesItemDecoration
+import com.coinbase.android.nativesdk.Wallet
+import com.google.android.material.bottomsheet.BottomSheetDialogFragment
+import kotlinx.coroutines.launch
+
+class WalletProviderBottomSheetFragment : BottomSheetDialogFragment() {
```
Comment (bangtoven) [inline android/example/src/main/java/com/coinbase/android/beta/ui/WalletProviderBottomSheetFragment.kt:19]: what about
```suggestion
class WalletPickerBottomSheetFragment : BottomSheetDialogFragment() {
```

Comment (bangtoven) [inline android/example/src/main/java/com/coinbase/android/beta/ui/WalletProviderBottomSheetFragment.kt]: how about
```suggestion
class WalletPickerBottomSheetFragment : BottomSheetDialogFragment() {
```

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt (hunk 9 above)
Comment (AMITGOELNY) [reply android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt:253]: Updated.  openitnent now referenced statically, and isn't a required param for getInstance.

File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/MobileWalletProvider.kt (hunk 10 above)
Comment (AMITGOELNY) [reply android/walletsdk/src/main/java/com/coinbase/android/nativesdk/MobileWalletProvider.kt]: Removed WalletProvider

File: android/example/src/main/java/com/coinbase/android/beta/ui/WalletProviderBottomSheetFragment.kt (hunk 11 above)
Comment (AMITGOELNY) [reply android/example/src/main/java/com/coinbase/android/beta/ui/WalletProviderBottomSheetFragment.kt]: Renamed

**Review by vishnumad — APPROVED** [review]
